### PR TITLE
Auto-create `WOODPECKER_AGENT_SECRET` secret

### LINKED ISSUE #207: Agent pod crashes with "agent could not auth: please provide an auth token"
Hi, the problem as as the title states. The Helm chart version is 1.5.0, the latest as of this moment. 

The agent pod crashes. Upon inspecting the the logs of the agent pod state:
```
{"level":"info","time":"2024-07-18T17:33:24Z","message":"log level: info"}
{"level":"info","time":"2024-07-18T17:33:24Z","message":"no agent config found at '/etc/woodpecker/agent.conf', start with defaults"}
{"level":"fatal","error":"rpc error: code = Unknown desc = agent could not auth: please provide a token","time":"2024-07-18T17:33:24Z","message":"error running agent"}
```

It looks like the secret isn't being loaded.

The following is my Helm for the agent section of `values.yml`:
```
agent:
  enabled: true
  env:
    WOODPECKER_BACKEND_K8S_NAMESPACE: [...]
    WOODPECKER_BACKEND_K8S_STORAGE_CLASS: "standard"
  persistence:
    storageClass: "standard"
    size: "10Gi"
  replicaCount: 1
```
Looking at the environment variable section of `kubectl describe`ing the agent:
```
    Environment Variables from:
      woodpecker-secret  Secret  Optional: false
    Environment:
      WOODPECKER_BACKEND:                      kubernetes
      WOODPECKER_BACKEND_K8S_NAMESPACE:        [...]
      WOODPECKER_BACKEND_K8S_POD_ANNOTATIONS:
      WOODPECKER_BACKEND_K8S_POD_LABELS:
      WOODPECKER_BACKEND_K8S_STORAGE_CLASS:    standard
      WOODPECKER_BACKEND_K8S_STORAGE_RWX:      true
      WOODPECKER_BACKEND_K8S_VOLUME_SIZE:      10G
      WOODPECKER_CONNECT_RETRY_COUNT:          1
```

I believe there should be a `WOODPECKER_AGENT_SECRET` environment variable, but I don't see one.

I may be configuring something incorrectly. Any help would be greatly appreciated.

### PR DESCRIPTION
**Pull request by pat-s**
While #144 took a first stab, the implementation does not work in practice for users as the secret is not generated automatically. Right now, only an empty k8s secret is created which must be filled manually with data containing `WOODPECKER_AGENT_SECRET`.

As #207 shows, users ideally expect the server-agent connection to work OOB.

This new approach

- creates a k8s secret called `woodpecker-default-agent-secret` which contains a valid `WOODPECKER_AGENT_SECRET` 
- Creation is controlled by a boolean flag `createAgentSecret` (default true for `server`), which can be set to `false` if users want to control the creation themselves

`woodpecker-default-agent-secret` will be read by the server and agent (if both are in the same namespace).
If the `agent` chart is deployed to a different namespace, users must configure the same secret on their own as we can't ensure the creation of identical secrets across namespaces.

fix #207

## TODO

- [x] Update README/instructions
- [x] Add tests

## Breaking

As this PR changes the way how secrets are created and removes the default `woodpecker-custom-secret` (as this is not in use anymore), this might break existing installations, hence the "breaking" label and major bump.

### CODE REVIEW

**Comment by pat-s** [comment]
@anbraten @xoxys RFR

**Comment by pat-s** [comment]
ping @xoxys

File: charts/woodpecker/README.md [hunk 1]
```diff
@@ -70,7 +70,7 @@ resource "helm_release" "woodpecker" {
 | agent.env.WOODPECKER_BACKEND_K8S_VOLUME_SIZE | string | `"10G"` |  |
 | agent.env.WOODPECKER_CONNECT_RETRY_COUNT | string | `"1"` |  |
 | agent.env.WOODPECKER_SERVER | string | `"woodpecker-server:9000"` | Add the environment variables for the agent component |
-| agent.extraSecretNamesForEnvFrom | list | `["woodpecker-secret"]` | Add extra secret that is contains environment variables |
+| agent.extraSecretNamesForEnvFrom | list | `["woodpecker-default-agent-secret"]` | Add extra secret that is contains environment variables |
```
Comment (xoxys) [inline charts/woodpecker/README.md:73]:
```suggestion
| agent.extraSecretNamesForEnvFrom | list | `["woodpecker-default-agent-secret"]` | Add extra secret that contains environment variables |
```

File: charts/woodpecker/charts/agent/README.md [hunk 2]
```diff
@@ -29,7 +29,7 @@ A Helm chart for the Woodpecker agent
 | env.WOODPECKER_BACKEND_K8S_VOLUME_SIZE | string | `"10G"` |  |
 | env.WOODPECKER_CONNECT_RETRY_COUNT | string | `"1"` |  |
 | env.WOODPECKER_SERVER | string | `"woodpecker-server:9000"` | Add the environment variables for the agent component |
-| extraSecretNamesForEnvFrom | list | `["woodpecker-secret"]` | Add extra secret that is contains environment variables |
+| extraSecretNamesForEnvFrom | list | `[]` | Add extra secret that is contains environment variables |
```
Comment (xoxys) [inline charts/woodpecker/charts/agent/README.md:32]:
```suggestion
| extraSecretNamesForEnvFrom | list | `[]` | Add extra secret that contains environment variables |
```

File: charts/woodpecker/charts/server/templates/secret.yaml [hunk 3]
```diff
@@ -11,3 +11,18 @@ data:
   {{ $key }}: {{ $value | b64enc }}
   {{- end }}
 {{- end }}
+---
+{{- if .Values.createAgentSecret }}
+{{- $secret := lookup "v1" "Secret" .Release.Namespace "woodpecker-default-agent-secret" -}}
+  {{- if or (not $secret) (not (hasKey $secret.data "WOODPECKER_AGENT_SECRET")) }}
+# Secret or key doesn't exist and createAgentSecret: true -> create secret
```
Comment (xoxys) [inline charts/woodpecker/charts/server/templates/secret.yaml:18]: IMO this indentation looks strange, and I would keep it at the same level as the wrapped content.

File: charts/woodpecker/charts/server/templates/secret.yaml [hunk 4]
```diff
@@ -11,3 +11,18 @@ data:
...
+apiVersion: v1
+kind: Secret
+metadata:
+  name: woodpecker-default-agent-secret
+  namespace: {{ .Release.Namespace }}
+type: Opaque
+data:
+  WOODPECKER_AGENT_SECRET: {{ randAlphaNum 64 | b64enc }}
```
Comment (xoxys) [inline charts/woodpecker/charts/server/templates/secret.yaml:26]: `randAlphaNum` should be secure enough, but was there a reason not to use `randAscii`?

File: charts/woodpecker/charts/server/values.yaml [hunk 5]
```diff
@@ -39,13 +39,15 @@ env:
 extraSecretNamesForEnvFrom:
   # - woodpecker-github-client
   # - woodpecker-github-secret
-  - woodpecker-secret
+  - woodpecker-default-agent-secret # created by `createAgentSecret: true`
```
Comment (xoxys) [inline charts/woodpecker/charts/server/values.yaml:42]: Does it make sense to rely on `extraSecretNamesForEnvFrom` here? If `createAgentSecret: true` is set, why not create and use the secret anyway, regardless if it is present in this list or not?

Comment (pat-s) [reply charts/woodpecker/charts/server/values.yaml:42]: Fair point! We can map it directly into the STS without relying on this variable.

**Review by xoxys — APPROVED** [review]

**Comment by pat-s** [comment]
@wrenix Thanks a lot, very valuable! And great to catch these before a release.
I didn't know about the `dig` approach so far, very handy!

#247

**Pull request merged**